### DISCUSSION TITLE
How do you Handle Model versioning ?

### DISCUSSION THREAD

**Discussion by anubhavshivamnath** [community/community#162357] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

How do you handle model versioning and experiment tracking in your ML pipelines? I'm looking for best practices to integrate MLflow with GitHub Actions for automated model deployment?

**Comment by AnubhavSC** [comment, marked as answer]
```yaml
# GitHub Workflow Trigger
on:
  push:
    branches:
      - main  # or a `release/` branch
```

### **Workflow Overview:**

1. **Trigger**: Code pushed to GitHub (`main` or `release` branch).
2. **Build Environment**: Set up Python, install deps.
3. **Run MLflow Training**: With tracking enabled.
4. **Log artifacts to MLflow**: Including model versions, metrics, tags.
5. **Check metrics threshold**: Automate model validation step.
6. **Promote model**: Register and move to "Staging" or "Production".
7. **Deploy**: Package with Docker or push to model serving infra (SageMaker, AzureML, local MLflow server, etc.).

---

## **Best Practices for MLflow Integration**

### ✅ **Use MLflow Tracking**

Track:

* `mlflow.log_param()` – learning rate, model type, etc.
* `mlflow.log_metric()` – accuracy, loss, etc.
* `mlflow.log_artifact()` – training plots, confusion matrix.
* `mlflow.sklearn.log_model()` – or `mlflow.pyfunc.log_model()` for general use.

### 📘 **Register and Manage Versions**

```python
result = mlflow.register_model(
    "runs:/<RUN_ID>/model", "my_model"
)
client.transition_model_version_stage(
    name="my_model",
    version=result.version,
    stage="Staging"
)
```

Automate this step in GitHub Actions if metrics meet threshold.

---

## . **GitHub Actions Integration**

Here’s a **minimal YAML** you can build on:

```yaml
name: Train and Deploy Model

on:
  push:
    branches:
      - main

jobs:
  train:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Set up Python
        uses: actions/setup-python@v4
        with:
          python-version: '3.10'

      - name: Install dependencies
        run: |
          pip install -r requirements.txt
          pip install mlflow boto3 scikit-learn  # etc.

      - name: Run training script
        env:
          MLFLOW_TRACKING_URI: ${{ secrets.MLFLOW_TRACKING_URI }}
        run: |
          python train.py
```

- Hope thsi will help you bud !
